### DISCUSSION TITLE
Help with analog audio streaming via adc

### DISCUSSION THREAD

**Discussion by mohaddeseh-rezaee** [pschatzmann/arduino-audio-tools#530] [Q&A] [answered]
Hi again. as you said before responding to discussion #493 I tried to see the simple streaming output. 
I gave a sin around 500Hz to ADC input and I see a sharp noise on it, no matter which frequency and voltage domain it is. I have checked my Doit-esp32-devkitv1 and read analog voltage via analogRead() and there is not this kind of noise. any recommend would be so much appreciated. 

```
#include "Arduino.h"
#include "BluetoothA2DPSource.h"
#include "AudioTools.h"

/**
 * @brief We use a mcp6022 analog microphone as input and send the data to A2DP
 */ 

AnalogAudioStream adc;
BluetoothA2DPSource a2dp_source;
/// @brief I added this part
const uint16_t sample_rate = 10000;
int channels = 2;
// The data has a center of around 26427, so we we need to shift it down to bring the center to 0
ConverterScaler<int16_t> scaler(1, 136, 2184);

// callback used by A2DP to provide the sound data
int32_t get_sound_data(Frame* frames, int32_t frameCount) {
    uint8_t *data = (uint8_t*)frames;
    int frameSize = 4;
    size_t resultBytes = adc.readBytes(data, frameCount*frameSize);   
    scaler.convert(data, resultBytes);
    return resultBytes/frameSize;
}

// Arduino Setup
void setup(void) {
  Serial.begin(115200);

  // start i2s input with default configuration
  Serial.println("starting I2S-ADC...");
  auto cfgRx = adc.defaultConfig(RX_MODE);
  cfgRx.sample_rate = sample_rate;
  cfgRx.channels = channels;
  adc.begin(cfgRx);
  //adc.begin(adc.defaultConfig(RX_MODE)); default string

  // start the bluetooth
  Serial.println("starting A2DP...");
  a2dp_source.set_auto_reconnect(true); 
  //static std::vector<char*> bt_names = {"B19","LG HBS760"};
  //a2dp_source.start(bt_names, get_sound_data);
  //a2dp_source.start(bt_names,get_sound_data);
  a2dp_source.start("B19",get_sound_data);
  
 }

// Arduino loop - repeated processing 
void loop() {
}
```

![stream](https://user-images.githubusercontent.com/119870635/208249787-163ab4be-a1aa-4668-8cc7-0220d6ac05a7.jpg)
![zoom](https://user-images.githubusercontent.com/119870635/208249795-c13b8d96-28b6-42b0-8b7f-b04efc32308e.jpg)

**Comment by pschatzmann** [comment, marked as answer]
Your sample rate is completely wrong, you need to provide the samples in the speed required by A2DP !

**Reply by mohaddeseh-rezaee** [reply]
Is 44100 ok for the sample rate? I have the same issue with a higher sample rate too. this is my export of data sampled via 44100 sample rate and given sin 495Hz 
![44100](https://user-images.githubusercontent.com/119870635/208251154-c85137d8-6d7b-4591-a66b-daa263066cd8.jpg)


[sin.txt](https://github.com/pschatzmann/arduino-audio-tools/files/10252117/sin.txt)

**Comment by pschatzmann** [comment]
A2DP only supports 44100

**Reply by mohaddeseh-rezaee** [reply]
I still have an issue with the default frequency of 44100. what shall I do?

**Comment by pschatzmann** [comment]
The only thing that I can think of is to fill a buffer first and then serve the data from the buffer. Life streaming is difficult and it is much easier to just send prerecorded audio e.g. from a file. 

I would run the output thru a resampler with an initial factor 1.0 and when you see that the buffer is running full or empty adust the re-sampling rate...

**Comment by mohaddeseh-rezaee** [comment]
I got clear output connecting both GPIO34 , GPIO14 and sampling rate of 44100. I need to know why it disconnects after about 35 seconds. If  it is because of buffering , is there any guidance to resolve it? 
my connection gets lost after  35 seconds and I don't know the source of problem. 
I am using esp32-devkit1  and using platformIO

**Comment by mohaddeseh-rezaee** [comment]
Phil, thanks so much for your time.
I got clear output connecting both GPIO34 , GPIO14 and sampling rate of 44100. I need to know why it disconnects after about 35 seconds. If it is because of buffering , is there any guidance to resolve it?
my connection gets lost after 35 seconds and I don't know the source of problem.
I am using esp32-devkit1 and using platformIO

**Comment by mohaddeseh-rezaee** [comment]
.

**Comment by pschatzmann** [comment]
My gut feeling is that A2DP notices that it does not get the data that is needed to play at a stable speed. The solution approach has been described above: you would need to link a PID controller to micro manage a Resampler is small steps to avoid any buffer over and underruns. 

You can try to increase I2S Sample Rate around 44100 a bit, to avoid buffer underruns, but in the end this will just lead to buffer over runs with some lost audio data... 

To reconfrm my assumption you could try out this sketch: https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-basic-api/base-player-a2dp/base-player-a2dp.ino

**Comment by mohaddeseh-rezaee** [comment]
![image](https://user-images.githubusercontent.com/119870635/209106091-b862f9b8-0fcd-4d46-99d6-9216ea778c1e.png)
